File: dotnet-desktop-guide/net/wpf/events/snippets/routed-events-overview/WpfControlLibraryCsharp/CustomButton.cs
using System.Windows;
using System.Windows.Controls;

namespace WpfControl
{
    public class CustomButton : Button
    {
        //<RegisterRouteEventAndClrAccessors>
        // Register a custom routed event using the Bubble routing strategy.
        public static readonly RoutedEvent TapEvent = EventManager.RegisterRoutedEvent(
            name: "Tap",
            routingStrategy: RoutingStrategy.Bubble,
            handlerType: typeof(RoutedEventHandler),
            ownerType: typeof(CustomButton));

        // Provide CLR accessors for adding and removing an event handler.
        public event RoutedEventHandler Tap
        {
            add { AddHandler(TapEvent, value); }
            remove { RemoveHandler(TapEvent, value); }
        }
        //</RegisterRouteEventAndClrAccessors>

        void RaiseCustomRoutedEvent()
        {
            // Create a RoutedEventArgs instance.
            RoutedEventArgs routedEventArgs = new(routedEvent: TapEvent);

            // Raise the event, which will bubble up through the element tree.
            RaiseEvent(routedEventArgs);
        }

        // For demo purposes, we use the Click event as a trigger.
        protected override void OnClick()
        {
            RaiseCustomRoutedEvent();

            // Call the base class OnClick() method so Click event subscribers are notified.
            base.OnClick();
        }
    }
}

File: dotnet-desktop-guide/wpf/events/snippets/routed-events-overview/WpfControlLibraryVb/CustomButton.vb
Imports System.Windows
Imports System.Windows.Controls

Public Class CustomButton
    Inherits Button

    '<RegisterRouteEventAndClrAccessors>
    ' Register a custom routed event using the Bubble routing strategy.
    Public Shared ReadOnly TapEvent As RoutedEvent = EventManager.RegisterRoutedEvent(
        name:="Tap",
        routingStrategy:=RoutingStrategy.Bubble,
        handlerType:=GetType(RoutedEventHandler),
        ownerType:=GetType(CustomButton))

    ' Provide CLR accessors for adding and removing an event handler.
    Public Custom Event Tap As RoutedEventHandler
        AddHandler(value As RoutedEventHandler)
            [AddHandler](TapEvent, value)
        End AddHandler

        RemoveHandler(value As RoutedEventHandler)
            [RemoveHandler](TapEvent, value)
        End RemoveHandler

        RaiseEvent(sender As Object, e As RoutedEventArgs)
            [RaiseEvent](e)
        End RaiseEvent
    End Event
    '</RegisterRouteEventAndClrAccessors>

    Private Sub RaiseCustomRoutedEvent()
        Dim routedEventArgs As New RoutedEventArgs(routedEvent:=TapEvent)
        [RaiseEvent](routedEventArgs)
    End Sub

    Protected Overrides Sub OnClick()
        RaiseCustomRoutedEvent()
        MyBase.OnClick()
    End Sub
End Class
